File: web_client/tests/forms/default_values.test.js
import { test, describe, it, before, beforeEach } from 'node:test';
import assert from 'node:assert';
import { JSDOM } from 'jsdom';

const dom = new JSDOM('<!DOCTYPE html><html><body></body></html>', { url: 'http://localhost/' });
global.window = dom.window;
global.document = dom.window.document;
global.HTMLElement = dom.window.HTMLElement;

const { RadioFieldBuilder } = await import('../../src/components/forms/fields/radio_field_builder.js');
const { CheckBoxFieldBuilder } = await import('../../src/components/forms/fields/check_box_field_builder.js');
const { OptionBuilderHelper } = await import('../../src/components/forms/fields/option_builder_helper.js');

const formModel = {
    relatedFields: [],
    occasionFeatures: [],
    occasionStartTime: null,
};

const makeOption = (title) => ({ title, id: null, price: 0, description: null });

describe('Default values — shouldPreselect', () => {
    it('radio: returns true when default_value matches option title', () => {
        const field = { data: { default_value: 'Option A' } };
        assert.strictEqual(
            OptionBuilderHelper.shouldPreselect(field, makeOption('Option A'), 'radio'),
            true
        );
    });

    it('radio: returns false when default_value does not match option title', () => {
        const field = { data: { default_value: 'Option A' } };
        assert.strictEqual(
            OptionBuilderHelper.shouldPreselect(field, makeOption('Option B'), 'radio'),
            false
        );
    });

    it('radio: returns false when default_value is missing', () => {
        const field = { data: {} };
        assert.strictEqual(
            OptionBuilderHelper.shouldPreselect(field, makeOption('Option A'), 'radio'),
            false
        );
    });

    it('radio: field.value takes precedence over default — returns false', () => {
        const field = { data: { default_value: 'Option A' }, value: 'Option B' };
        assert.strictEqual(
            OptionBuilderHelper.shouldPreselect(field, makeOption('Option A'), 'radio'),
            false
        );
    });

    it('checkbox: returns true when default_value array contains option title', () => {
        const field = { data: { default_value: ['Option A', 'Option C'] } };
        assert.strictEqual(
            OptionBuilderHelper.shouldPreselect(field, makeOption('Option A'), 'checkbox'),
            true
        );
        assert.strictEqual(
            OptionBuilderHelper.shouldPreselect(field, makeOption('Option C'), 'checkbox'),
            true
        );
    });

    it('checkbox: returns false when option title is not in default_value array', () => {
        const field = { data: { default_value: ['Option A'] } };
        assert.strictEqual(
            OptionBuilderHelper.shouldPreselect(field, makeOption('Option B'), 'checkbox'),
            false
        );
    });

    it('checkbox: returns false when default_value is a non-array (radio default applied to checkbox field)', () => {
        const field = { data: { default_value: 'Option A' } };
        assert.strictEqual(
            OptionBuilderHelper.shouldPreselect(field, makeOption('Option A'), 'checkbox'),
            false
        );
    });
});

describe('Default values — RadioFieldBuilder.create', () => {
    it('pre-selects radio matching default_value', () => {
        const field = {
            id: 'fld1',
            type: 'select_one',
            title: 'Choose',
            options: [makeOption('Yes'), makeOption('No'), makeOption('Maybe')],
            data: { default_value: 'No' },
        };
        const el = RadioFieldBuilder.create(field, formModel);
        const inputs = el.querySelectorAll('input[type=radio]');
        const checked = Array.from(inputs).filter(i => i.checked).map(i => i.value);
        assert.deepStrictEqual(checked, ['No']);
    });

    it('does not pre-select anything when default points to deleted option', () => {
        const field = {
            id: 'fld2',
            type: 'select_one',
            title: 'Choose',
            options: [makeOption('Yes'), makeOption('No')],
            data: { default_value: 'Deleted Option' },
        };
        const el = RadioFieldBuilder.create(field, formModel);
        const checkedCount = el.querySelectorAll('input[type=radio]:checked').length;
        assert.strictEqual(checkedCount, 0);
    });

    it('does not pre-select anything when no default is set', () => {
        const field = {
            id: 'fld3',
            type: 'select_one',
            title: 'Choose',
            options: [makeOption('Yes'), makeOption('No')],
            data: {},
        };
        const el = RadioFieldBuilder.create(field, formModel);
        const checkedCount = el.querySelectorAll('input[type=radio]:checked').length;
        assert.strictEqual(checkedCount, 0);
    });

    it('field.value (existing answer) overrides default', () => {
        const field = {
            id: 'fld4',
            type: 'select_one',
            title: 'Choose',
            options: [makeOption('Yes'), makeOption('No')],
            data: { default_value: 'Yes' },
            value: 'No',
        };
        const el = RadioFieldBuilder.create(field, formModel);
        const checked = Array.from(el.querySelectorAll('input[type=radio]'))
            .filter(i => i.checked).map(i => i.value);
        assert.deepStrictEqual(checked, ['No']);
    });
});

describe('Default values — CheckBoxFieldBuilder.create', () => {
    it('pre-checks all options matching default_value array', () => {
        const field = {
            id: 'fld5',
            type: 'select_many',
            title: 'Pick many',
            options: [makeOption('A'), makeOption('B'), makeOption('C')],
            data: { default_value: ['A', 'C'] },
        };
        const el = CheckBoxFieldBuilder.create(field, formModel);
        const checked = Array.from(el.querySelectorAll('input[type=checkbox]'))
            .filter(i => i.checked).map(i => i.value);
        assert.deepStrictEqual(checked.sort(), ['A', 'C']);
    });

    it('ignores defaults that refer to deleted options', () => {
        const field = {
            id: 'fld6',
            type: 'select_many',
            title: 'Pick many',
            options: [makeOption('A'), makeOption('B')],
            data: { default_value: ['A', 'Deleted'] },
        };
        const el = CheckBoxFieldBuilder.create(field, formModel);
        const checked = Array.from(el.querySelectorAll('input[type=checkbox]'))
            .filter(i => i.checked).map(i => i.value);
        assert.deepStrictEqual(checked, ['A']);
    });

    it('pre-checks nothing when no default is set', () => {
        const field = {
            id: 'fld7',
            type: 'select_many',
            title: 'Pick many',
            options: [makeOption('A'), makeOption('B')],
            data: {},
        };
        const el = CheckBoxFieldBuilder.create(field, formModel);
        const checkedCount = el.querySelectorAll('input[type=checkbox]:checked').length;
        assert.strictEqual(checkedCount, 0);
    });
});

File: test/components/forms/default_values_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:fstapp/components/forms/models/form_option_model.dart';
import 'package:fstapp/components/forms/models/holder_models/field_holder.dart';
import 'package:fstapp/components/forms/widgets_view/check_box_field_builder.dart';
import 'package:fstapp/components/forms/widgets_view/form_helper.dart';
import 'package:fstapp/components/forms/widgets_view/radio_field_builder.dart';

void main() {
  group('RadioFieldBuilder.resolveDefault (select_one)', () {
    final options = [
      FormOptionModel('0', 'Yes'),
      FormOptionModel('1', 'No'),
      FormOptionModel('2', 'Maybe'),
    ];

    OptionsFieldHolder holderWith(dynamic defaultValue) => OptionsFieldHolder(
          id: 1,
          fieldType: FormHelper.fieldTypeSelectOne,
          isRequired: false,
          options: options,
          value: defaultValue,
          title: 'Q',
        );

    test('returns matching FormOptionModel when default matches option title',
        () {
      final resolved = RadioFieldBuilder.resolveDefault(holderWith('No'), options);
      expect(resolved, isNotNull);
      expect(resolved!.title, 'No');
    });

    test('returns null when default points to a deleted option', () {
      final resolved = RadioFieldBuilder.resolveDefault(
          holderWith('Gone'), options);
      expect(resolved, isNull);
    });

    test('returns null when default is empty string', () {
      final resolved = RadioFieldBuilder.resolveDefault(holderWith(''), options);
      expect(resolved, isNull);
    });

    test('returns null when default is missing', () {
      final resolved = RadioFieldBuilder.resolveDefault(holderWith(null), options);
      expect(resolved, isNull);
    });

    test('returns null when default is wrong type (list instead of string)', () {
      final resolved = RadioFieldBuilder.resolveDefault(
          holderWith(['No']), options);
      expect(resolved, isNull);
    });
  });

  group('CheckboxFieldBuilder.resolveDefaults (select_many)', () {
    final options = [
      FormOptionModel('0', 'A'),
      FormOptionModel('1', 'B'),
      FormOptionModel('2', 'C'),
    ];

    OptionsFieldHolder holderWith(dynamic defaultValue) => OptionsFieldHolder(
          id: 2,
          fieldType: FormHelper.fieldTypeSelectMany,
          isRequired: false,
          options: options,
          value: defaultValue,
          title: 'Q',
        );

    test('returns matching options when all defaults exist', () {
      final resolved = CheckboxFieldBuilder.resolveDefaults(
          holderWith(['A', 'C']), options);
      expect(resolved.map((o) => o.title).toList()..sort(), ['A', 'C']);
    });

    test('silently drops defaults that refer to deleted options', () {
      final resolved = CheckboxFieldBuilder.resolveDefaults(
          holderWith(['A', 'Gone']), options);
      expect(resolved.map((o) => o.title).toList(), ['A']);
    });

    test('returns empty list when default is missing', () {
      final resolved = CheckboxFieldBuilder.resolveDefaults(
          holderWith(null), options);
      expect(resolved, isEmpty);
    });

    test('returns empty list when default is wrong type (string instead of list)',
        () {
      final resolved = CheckboxFieldBuilder.resolveDefaults(
          holderWith('A'), options);
      expect(resolved, isEmpty);
    });

    test('returns empty list when default list is empty', () {
      final resolved =
          CheckboxFieldBuilder.resolveDefaults(holderWith([]), options);
      expect(resolved, isEmpty);
    });

    test('ignores non-string entries in default list', () {
      final resolved = CheckboxFieldBuilder.resolveDefaults(
          holderWith(['A', 42, null, 'B']), options);
      expect(resolved.map((o) => o.title).toList()..sort(), ['A', 'B']);
    });
  });
}
